File: scala-master-12/src/main/scala/usecase/OverLapSolution.scala
package usecase

import org.apache.spark.SparkContext
import scala.io.Source
import java.io.{ FileReader, FileNotFoundException, IOException }
import org.apache.spark.sql.SQLContext
import org.apache.spark.sql.Row
import org.apache.spark.sql.types.{ StructType, StructField, StringType, IntegerType };
import scala.collection.mutable.ListBuffer
import org.apache.spark.sql.SparkSession
import org.apache.hadoop.log.LogLevel
import org.apache.spark.sql.functions._
import org.apache.spark.sql.SaveMode

object OverLapSolution {
  
  def main(args: Array[String]): Unit = {
    val zipCodeRangeList : List[ZipCodeRange]  = generateData()
    val normalMap : scala.collection.mutable.SortedMap[Long,ZipCodeRange] = scala.collection.mutable.SortedMap.empty[Long,ZipCodeRange]
    val controlMap : scala.collection.mutable.SortedMap[Long,ZipCodeRange] = scala.collection.mutable.SortedMap.empty[Long,ZipCodeRange](normalMap.ordering.reverse)

    
    val outputZipList : scala.collection.mutable.Set[ZipCodeRange] = scala.collection.mutable.Set.empty[ZipCodeRange]
    for(zipCodeRange <- zipCodeRangeList.iterator){
      controlMap += ((String.valueOf(zipCodeRange.endZip-zipCodeRange.startZip)+String.valueOf(System.currentTimeMillis())).toLong -> zipCodeRange)
    }

    controlMap.foreach(f => {
      if(controlMap.contains(f._1)){
           val superStart : Long = controlMap.get(f._1).get.startZip
           val superEnd : Long = controlMap.get(f._1).get.endZip
           val rangeList : scala.collection.mutable.Set[Long] = scala.collection.mutable.Set.empty[Long]
           rangeList += superStart 
           rangeList += superEnd
           controlMap.foreach(x => {
            if(controlMap.contains(x._1)){ 
              val subStart : Long = controlMap.get(x._1).get.startZip
              val subEnd : Long = controlMap.get(x._1).get.endZip
              if(subEnd >= superEnd && subStart <= superEnd) {
                          rangeList += subEnd
                          controlMap -= (x._1)
                      }else if(subStart <= superStart && subEnd >= superStart) {
                          rangeList += subStart
                          controlMap -= (x._1)
                      }else if(subStart >= superStart && subEnd <= superEnd) {
                          controlMap -= (x._1)
                      }else{}
            }  
           })           
           outputZipList += new ZipCodeRange(rangeList.min,rangeList.max)
      }
    })    
    outputZipList.foreach(f => println(f.startZip+" - "+f.endZip))        
  }

  
  def generateData(): List[ZipCodeRange] = {
    val zipCodeRangeList : scala.collection.mutable.ListBuffer[ZipCodeRange] = scala.collection.mutable.ListBuffer.empty[ZipCodeRange]
    // add element
    zipCodeRangeList.toList
  }
  
}

File: scala-master-12/src/main/java/usecase/OverLapSolution.java
package usecase;

import java.util.concurrent.ConcurrentSkipListMap;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Comparator;
import java.util.Map.Entry;
import java.util.Set;
import java.util.TreeSet;

public class OverLapSolution {

	public static void main(String[] args) {
		
	    List <ZipCodeRangePojo> inputZipList = generateData();
	    List <ZipCodeRangePojo> outputZipList = new ArrayList<ZipCodeRangePojo>();
	    ConcurrentSkipListMap<Long,ZipCodeRangePojo> controlMap = new ConcurrentSkipListMap<Long,ZipCodeRangePojo>(Comparator.reverseOrder());
	    for(ZipCodeRangePojo zipCodeRange : inputZipList) {
	    	controlMap.put(Long.valueOf(String.valueOf(zipCodeRange.endZip-zipCodeRange.startZip)+String.valueOf(System.currentTimeMillis())),zipCodeRange);
	    }
	    
	    
		for(Entry<Long,ZipCodeRangePojo> superDataSet : controlMap.entrySet()) {
		     if(controlMap.containsKey(superDataSet.getKey())){
		        long superStart = controlMap.get(superDataSet.getKey()).startZip;
		        long superEnd = controlMap.get(superDataSet.getKey()).endZip;
		        Set<Long> rangeList = new TreeSet<Long>();
		        rangeList.add(superStart);rangeList.add(superEnd);

		            for(Entry<Long,ZipCodeRangePojo> subDataSet : controlMap.entrySet()) {
		              if(!superDataSet.getKey().equals(subDataSet.getKey())) {
		                  long subStart = controlMap.get(subDataSet.getKey()).startZip;
		                  long subEnd = controlMap.get(subDataSet.getKey()).endZip;
		                  // Check the Range
		                  if(subEnd >= superEnd && subStart <= superEnd) {
		                      rangeList.add(subEnd);
		                      controlMap.remove(subDataSet.getKey());
		                  }else if(subStart <= superStart && subEnd >= superStart) {
		                      rangeList.add(subStart);
		                      controlMap.remove(subDataSet.getKey());
		                  }else if(subStart >= superStart && subEnd <= superEnd) {
		                      controlMap.remove(subDataSet.getKey());
		                  }else
		                      ;
		                  System.out.println("Size of controlMap "+controlMap.size());
		              }
		            }
		            outputZipList.add(new ZipCodeRangePojo(Collections.min(rangeList),Collections.max(rangeList)));
	}
	// Final ranges are
		     for(ZipCodeRangePojo out : outputZipList) {
		    	 System.out.println(out.startZip+"  -- "+out.endZip);
		     }
 }
}

	private static List<ZipCodeRangePojo> generateData(){
		List<ZipCodeRangePojo> l = new ArrayList<ZipCodeRangePojo>();
		//....
		return l;
	}
	
}	
